src/com/facebook/buck/jvm/java/JavacPipelineState.java

```diff
@@ -125,25 +127,37 @@ public boolean isRunning() {
                   .map(ResolvedJavacPluginProperties::getJavacPluginJsr199Fields)
                   .collect(Collectors.toList()));
 
-      invocation =
-          getJavac()
-              .newBuildInvocation(
-                  javacExecutionContext,
-                  resolver,
-                  invokingRule,
-                  getOptions(context, compilerParameters.getClasspathEntries()),
-                  pluginFields,
-                  compilerParameters.getSourceFilePaths(),
-                  compilerParameters.getPathToSourcesList(),
-                  compilerParameters.getWorkingDirectory(),
-                  compilerParameters.shouldTrackClassUsage(),
-                  abiJarParameters,
-                  libraryJarParameters,
-                  compilerParameters.getAbiGenerationMode(),
-                  compilerParameters.getAbiCompatibilityMode(),
-                  compilerParameters.getSourceOnlyAbiRuleInfo());
-
-      closeables.add(invocation);
+      /* When compiling a kotlin project, we can't know if there will or not have annotation
+         processors that generates java source files to compile. So we pass the directories
+         where potencially generated files will be, and here, which runs after all kotlin steps
+         we filter for only java files, or don't invoke the compiler, if there aren't any */
```

Comment:
This is another reason to do these as rules rather than steps -- you won't need to special case this since a `java_library` rule with empty sources already does the right thing.